bittide/src/Bittide/Wishbone.hs

```diff
@@ -111,3 +117,165 @@ singleMasterInterconnect' config master slaves = (toMaster, bundle toSlaves)
     case divWithRemainder @(Regs a 8) @8 @7 of
       Dict ->
         f (master, unbundle slaves)
+
+unsafeToDf :: Circuit (CSignal dom (Maybe a)) (Df dom a)
+unsafeToDf = Circuit $ \ (CSignal cSig, _) -> (CSignal $ pure (), maybeToData <$> cSig)
+
+uartDf ::
+  (HiddenClockResetEnable dom, ValidBaud dom baud) =>
+  SNat baud ->
+  Circuit
+    (Df dom (BitVector 8), CSignal dom Bit)
+    (CSignal dom (Maybe (BitVector 8)), CSignal dom Bit)
+uartDf baud = Circuit go
+ where
+  go ((request, ~(CSignal rxBit)),_) = ((Ack <$> ack, CSignal $ pure ()), (CSignal received, CSignal txBit))
+   where
+    (received, txBit, ack) = uart baud rxBit (dataToMaybe <$> request)
+
+uartWb ::
+  forall dom addrW nBytes baudRate transmitBufferDepth receiveBufferDepth .
+  ( HiddenClockResetEnable dom, ValidBaud dom baudRate
+  , 1 <= transmitBufferDepth
+  , 1 <= receiveBufferDepth
+  , 2 <= addrW
+  , KnownNat addrW
+  , KnownNat nBytes, 1 <= nBytes
+  ) =>
+  SNat transmitBufferDepth ->
+  SNat receiveBufferDepth ->
+  SNat baudRate ->
+  Circuit
+    (Wishbone dom 'Standard addrW (Bytes nBytes), CSignal dom Bit)
+    (CSignal dom Bit, CSignal dom (Bool, Bool))
+uartWb txDepth@SNat rxDepth@SNat baud = circuit $ \(wb, uartRx) -> do
+  (txFifoIn, uartStatus) <- wbToDf -< (wb, rxFifoOut, txFifoMeta)
+  (txFifoOut, txFifoMeta) <- fifoWithMeta txDepth -< txFifoIn
+  (rxFifoIn, uartTx) <- uartDf baud -< (txFifoOut, uartRx)
+  (rxFifoOut, _rx') <- fifoWithMeta rxDepth <| unsafeToDf -< rxFifoIn
+  idC -< (uartTx, uartStatus)
+
+wbToDf ::
+  forall dom addrW nBytes txFifoDepth .
+  ( KnownNat nBytes, 1 <= nBytes, KnownNat addrW, 2 <= addrW) =>
+  Circuit
+    (Wishbone dom 'Standard addrW (Bytes nBytes), Df dom (BitVector 8), CSignal dom (FifoMeta txFifoDepth))
+    (Df dom (BitVector 8), CSignal dom (Bool, Bool))
+wbToDf = Circuit $
+  bimap (third CSignal . unbundle) (second CSignal . unbundle) .
+  unbundle .
+  fmap go .
+  bundle .
+  bimap (bundle . third unCSignal) (bundle . second unCSignal)
+ where
+  third f (a, b, c) = (a, b, f c)
+  unCSignal (CSignal s) = s
+  go ((WishboneM2S{..}, dataToMaybe -> rxData, fifoFull -> txFull), (Ack txAck, _)) =
+    ( ((emptyWishboneS2M @()){acknowledge, err, readData}, Ack validRead, ())
+    , (txWrite, status)
+    )
+   where
+    (alignedAddr, alignment) = split @_ @(addrW - 2) @2 addr
+    internalAddr = bitCoerce $ resize alignedAddr :: Index 2
+    addrLegal = alignedAddr <= 1 && alignment == 0
+    masterActive = busCycle && strobe
+    validRead = masterActive && internalAddr == 0 && not writeEnable
+    validWrite = masterActive && internalAddr == 0 && writeEnable && txAck
+    validReadStatus = masterActive && internalAddr == 1 && not writeEnable
+
+    acknowledge = (validRead || validWrite || validReadStatus) && addrLegal
+    err = masterActive && (not (validRead || validWrite || validReadStatus) || not addrLegal)
```

Comment:
I personally find the flow here pretty hard to read. The value of `ack` or `err` depend on a lot of different things that are set somewhere else. What do you think about using guard-clauses to split all those different cases into separate function bodies?

```hs
-- not in cycle
| not (busCycle && strobe) = ((emptyWishboneS2M, Ack False, ()), (NoData, status))
-- illegal addr
| not addrLegal = ((emptyWishboneS2M { err = True }, Ack False, ()), (NoData, status)
-- read at 0
| not writeEnable && internalAddr == 0 =
    ( (emptyWishboneS2M { ack = True, readData = resize $ fromMaybe 0 rxData }, Ack True, ()),
      (NoData, status))
-- write at 0
| writeEnable && internalAddr == 0 =
    ( (emptyWishboneS2M { ack = True, readData = invalidReq }, Ack False, ()),
      (Data $ resize writeData, status))
-- read at 1
| not writeEnable && internalAddr == 1 =
    ( (emptyWishboneS2M { ack = True, readData = resize $ pack status }, Ack False, (),
      (NoData, status))
 
| otherwise = ((emptyWishboneS2M { err = True }, Ack False, ()), (NoData, status))
```

Reply:
I agree that this is nicer, I changed it